src/irr/asset/CGLSLBRDFBuiltinIncludeLoader.h

```diff
@@ -1,18 +1,209 @@
-#ifndef __IRR_C_GLSL_BRDF_BUILTIN_INCLUDE_LOADER_H_INCLUDED__
-#define __IRR_C_GLSL_BRDF_BUILTIN_INCLUDE_LOADER_H_INCLUDED__
+#ifndef __IRR_C_GLSL_BSDF_BUILTIN_INCLUDE_LOADER_H_INCLUDED__
+#define __IRR_C_GLSL_BSDF_BUILTIN_INCLUDE_LOADER_H_INCLUDED__
+
+//TODO this file should change name to CGLSLBSDFBuiltinIncludeLoader.h
 
 #include "irr/asset/IBuiltinIncludeLoader.h"
 
 namespace irr {
 namespace asset
 {    
 
-class CGLSLBRDFBuiltinIncludeLoader : public irr::asset::IBuiltinIncludeLoader
+class CGLSLBSDFBuiltinIncludeLoader : public irr::asset::IBuiltinIncludeLoader
 {
 public:
-    const char* getVirtualDirectoryName() const override { return "glsl/brdf/"; }
+    const char* getVirtualDirectoryName() const override { return "glsl/bsdf/"; }
 
 private:
+    static std::string getCommons(const std::string&)
+    {
+        return
+R"(#ifndef _IRR_BSDF_COMMON_INCLUDED_
+#define _IRR_BSDF_COMMON_INCLUDED_
+
+// do not use this struct in SSBO or UBO, its wasteful on memory
+struct irr_glsl_DirAndDifferential
+{
+   vec3 dir;
+   // differentials at origin, I'd much prefer them to be differentials of barycentrics instead of position in the future
+   mat2x3 dPosdScreen;
+};
+
+// do not use this struct in SSBO or UBO, its wasteful on memory
+struct irr_glsl_ViewSurfaceInteraction
+{
+   irr_glsl_DirAndDifferential V; // outgoing direction, NOT NORMALIZED; V.dir can have undef value for lambertian BSDF
+   vec3 N; // surface normal, NOT NORMALIZED
+};
+
+// do not use this struct in SSBO or UBO, its wasteful on memory
+struct irr_glsl_BSDFSample
+{
+   vec3 L;  // incoming direction, normalized
+   float probability; // for a single sample (don't care about number drawn)
+};
+
+
+// do not use this struct in SSBO or UBO, its wasteful on memory
+struct irr_glsl_BSDFIsotropicParams
+{
+   float NdotL;
+   float NdotL_squared;
+   float NdotV;
+   float NdotV_squared;
+   float VdotL; // same as LdotV
+   float NdotH;
+   float VdotH; // same as LdotH
+   // left over for anisotropic calc and BSDF that want to implement fast bump mapping
+   float LplusV_rcpLen;
+   // basically metadata
+   vec3 L;
+   irr_glsl_ViewSurfaceInteraction interaction;
+};
+
+// do not use this struct in SSBO or UBO, its wasteful on memory
+struct irr_glsl_BSDFAnisotropicParams
+{
+   irr_glsl_BSDFIsotropicParams isotropic;
+   float TdotL;
+   float TdotV;
+   float TdotH;
+   float BdotL;
+   float BdotV;
+   float BdotH;
+   // useless metadata
+   vec3 T;
+   vec3 B;
+};
+
+/*
+//TODO error no operator '*' for vec3 and mat2x3
+// chain rule on various functions (usually vertex attributes and barycentrics)
+vec2 irr_glsl_applyScreenSpaceChainRule1D3(in vec3 dFdG, in mat2x3 dGdScreen)
+{
+   return dFdG*dGdScreen;
```

Comment:
treat `dFdG` like a mat1x3, so `dFdG*dGdScreen` needs to change to a `vec2(dot(dFdG,dGdScreen[0]),dot(dFdG,dGdScreen[1]))`